langfuse/callback/llama_index.py

```diff
@@ -0,0 +1,366 @@
+            self.trace = self.langfuse.trace(
+                id=str(uuid4()),
+                name=self.trace_name or f"LlamaIndex_{self._cur_trace_id}",
+                version=self.version,
+                session_id=self.session_id,
+                user_id=self.user_id,
+            )
+
+            return self.trace
+
+    def _create_observation(
+        self,
+        event_id: str,
+        parent: Union[
+            StatefulTraceClient, StatefulSpanClient, StatefulGenerationClient
+        ],
+        trace_id: str,
+    ) -> Union[StatefulSpanClient, StatefulGenerationClient]:
+        event_type = self.event_map[event_id][0].event_type
+
+        if event_type == CBEventType.LLM:
+            return self._handle_LLM_events(event_id, parent, trace_id)
+        elif event_type == CBEventType.EMBEDDING:
+            return self._handle_embedding_events(event_id, parent, trace_id)
+        else:
+            return self._handle_span_events(event_id, parent, trace_id)
+
+    def _handle_LLM_events(
+        self,
+        event_id: str,
+        parent: Union[
+            StatefulTraceClient, StatefulSpanClient, StatefulGenerationClient
+        ],
+        trace_id: str,
+    ) -> StatefulGenerationClient:
+        start_event, end_event = self.event_map[event_id]
+
+        if start_event.payload and EventPayload.SERIALIZED in start_event.payload:
+            serialized = start_event.payload.get(EventPayload.SERIALIZED, {})
+            name = serialized.get("class_name", "LLM")
+            temperature = serialized.get("temperature", None)
+            max_tokens = serialized.get("max_tokens", None)
+            timeout = serialized.get("timeout", None)
+
+        if end_event.payload:
+            if EventPayload.PROMPT in end_event.payload:
+                input = end_event.payload.get(EventPayload.PROMPT)
+                output = end_event.payload.get(EventPayload.COMPLETION)
+
+            elif EventPayload.MESSAGES in end_event.payload:
+                input = end_event.payload.get(EventPayload.MESSAGES)
+                response = end_event.payload.get(EventPayload.RESPONSE, {})
+                output = response.message
+                model = response.raw.get("model", None)
+                token_usage = dict(response.raw.get("usage", {}))
+                usage = None
+                if token_usage:
+                    usage = {
+                        "prompt_tokens": token_usage.get("prompt_tokens"),
+                        "completion_tokens": token_usage.get("completion_tokens"),
+                        "total_tokens": token_usage.get("total_tokens"),
+                    }
+
+        generation = parent.generation(
+            id=event_id,
+            trace_id=trace_id,
+            version=self.version,
+            name=name,
+            start_time=start_event.time,
+            end_time=end_event.time,
+            usage=usage or None,
+            model=model,
+            input=input,
```

Comment:
chat format

Reply:
Check if additional_kwargs is filled on input or output, then send it, otherwise do not send it.